[git-webkit] Add land command

JonWBedard opened the pull request:
#### 2c52e40530cd68d794401010e06b0fb8449a72ce
<pre>
[git-webkit] Add land command
<a href="https://bugs.webkit.org/show_bug.cgi?id=231821">https://bugs.webkit.org/show_bug.cgi?id=231821</a>
&lt;rdar://problem/84309339 &gt;

Reviewed by NOBODY (OOPS!).

* Tools/Scripts/git-webkit: WebKit still uses Svn as it&apos;s canonical source of truth.
* Tools/Scripts/libraries/webkitscmpy/setup.py: Bump version.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/__init__.py: Ditto.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/local/git.py:
(Git._to_git_ref): Convert identifiers or revisions to git refs.
(Git.checkout): Use shared _to_git_ref function.
(Git.rebase): Support rebasing identifiers.
(Git.pull): Generalize computation of the gmtoffset.
(Git.diff_lines): Output diff between two refs.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/mocks/local/git.py:
(Git.__init__): Add branch, push and diff commands.
(Git.filter_branch): Support sed.
(Git.pull): Mock `git pull`
(Git.move_branch): Mock `git rebase`.
(Git.push): Mock `git push`.
(Git.reset): Mock `git reset HEAD~#`.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/__init__.py:
(main): Add land command.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/land.py: Added.
(Land.parser):
(Land.main): Rebase the current branch against it&apos;s root, assign the reviewer, check for &apos;OOPS&apos; messages,
rebase against target branch, update target branch ref, optionally canonicalize, push (or svn dcommit) and
update and close pull-request with information on the landed commit.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/scm_base.py:
(ScmBase.gmtoffset): Add shared computation of GMT offset.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/test/git_unittest.py:
(TestGit.test_commits): `git log` should exclude base ref.
(TestGit.test_commits_branch): Ditto.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/test/land_unittest.py: Added.
(repository):
(TestLand):
(TestLandGitHub):
(TestLandBitBucket):
</pre>

Gyuyoung commented:
Land without review.

dewei-zhu commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/local/git.py:
```diff
@@ -912,3 +916,27 @@ def modified(self, staged=None):
+            cwd=self.root_path,
+            stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE,
+            ** kwargs
```
Nit: do we expect space between ** and kwargs?

dewei-zhu replied:
Nit: no space between ** and kwargs(?)

dewei-zhu commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/local/git.py:
```diff
@@ -912,3 +916,27 @@ def modified(self, staged=None):
+        head = self._to_git_ref(head)
+
+        kwargs = dict()
+        if sys.version_info >= (3, 0):
```
`sys.version_info >= (3, 6)`?

dewei-zhu replied:
`sys.version_info >= (3, 6)` ?

JonWBedard committed:
[git-webkit] Add land command
https://bugs.webkit.org/show_bug.cgi?id=231821
<rdar://problem/84309339>

Reviewed by NOBODY (OOPS!).

* Tools/Scripts/git-webkit: WebKit still uses Svn as it's canonical source of truth.
* Tools/Scripts/libraries/webkitscmpy/setup.py: Bump version.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/__init__.py: Ditto.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/local/git.py:
(Git._to_git_ref): Convert identifiers or revisions to git refs.
(Git.checkout): Use shared _to_git_ref function.
(Git.rebase): Support rebasing identifiers.
(Git.pull): Generalize computation of the gmtoffset.
(Git.diff_lines): Output diff between two refs.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/mocks/local/git.py:
(Git.__init__): Add branch, push and diff commands.
(Git.filter_branch): Support sed.
(Git.pull): Mock `git pull`
(Git.move_branch): Mock `git rebase`.
(Git.push): Mock `git push`.
(Git.reset): Mock `git reset HEAD~#`.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/__init__.py:
(main): Add land command.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/land.py: Added.
(Land.parser):
(Land.main): Rebase the current branch against it's root, assign the reviewer, check for 'OOPS' messages,
rebase against target branch, update target branch ref, optionally canonicalize, push (or svn dcommit) and
update and close pull-request with information on the landed commit.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/scm_base.py:
(ScmBase.gmtoffset): Add shared computation of GMT offset.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/test/git_unittest.py:
(TestGit.test_commits): `git log` should exclude base ref.
(TestGit.test_commits_branch): Ditto.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/test/land_unittest.py: Added.
(repository):
(TestLand):
(TestLandGitHub):
(TestLandBitBucket):
